### ISSUE TITLE
Add an `-Until` script block parameter to `Select-Object`

### ISSUE THREAD

**Issue by MatejKafka** [PowerShell/PowerShell#21610]
### Summary of the new feature / enhancement

As a user, I want to stop processing input after encountering a specific value in the pipeline.

A simple example:
```powershell
$NumbersUnder50 = 0..1000 | select -Until {$_ -eq 50} # stop after reaching 50
```

A more realistic example:
```powershell
# list GitHub releases until a specific version
# with -FollowRelLink, this may result in saving a lot of HTTP requests under the hood, since otherwise all releases are listed
Invoke-RestMethod -FollowRelLink "https://api.github.com/repos/PowerShell/PowerShell/releases" `
    | % {$_} | select -Until {$_.tag_name -eq "v7.2.0"}
```

Since `StopUpstreamCommandsException` , which is internally used by `Select-Object` to implement its early-exit behavior is internal, I don't believe there's currently any way to implement this without reflecting PowerShell internals. Using something like `... | % {if (...) {break}; $_}` may seem like an attractive option, but [as documented](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_break?view=powershell-7.4), it terminates much more (the whole runspace) than a user would naively expect.

A more general solution could be to make `StopUpstreamCommandsException` public, so that users can write their own cmdlets with support for early exit, which is one of the possible solutions for #3821. However, even if that gets implemented, I believe the `-Until` parameter could still be valuable for simpler use cases.

### Proposed technical implementation details (optional)

Add an `-Until` script block parameter to `Select-Object`, probably in a new parameter set, run the passed script block for each pipeline item and throw `StopUpstreamCommandsException` when the script block returns a truthy value.

I can open a PR if the feature is accepted. (although I'd kinda appreciate if someone found time to review my other PRs first, nudge nudge wink wink :) )

**Comment by SteveL-MSFT** [comment]
I wonder if it would be more generally useful to have a `Stop-Pipeline` cmdlet that emits the internal exception?

**Comment by MatejKafka** [comment]
@SteveL-MSFT It would definitely be more generally useful. However, there's already an open issue for that, and there doesn't seem to be will or time to design and implement that feature, and this might be a stepping stone that's easier to decide on and implement, and should be a useful addition on its own.

As an aside, I'd preferably make the exception public instead of creating a `Stop-Pipeline` cmdlet, so that it can be easily invoked from compiled cmdlets, not just from scripts.

**Comment by SteveL-MSFT** [comment]
@MatejKafka can you link to the other issue on this?  I've queued this up for the WG-Cmdlets to discuss so a conclusion will be made

**Comment by mklement0** [comment]
@SteveL-MSFT, said issue is already linked to from the initial post:

* #3821.

**Comment by SteveL-MSFT** [comment]
@PowerShell/wg-powershell-cmdlets discussed this.  There are existing scripting alternatives to solve this problem and we don't believe we want to keep adding capability to `Select-Object`.  Here's an example function that can be used:

```powershell
#requires -Version 5.1
Function Select-Until {
    [cmdletbinding()]
    [alias('su')]
    Param(
        [Parameter(
            Position = 0,
            Mandatory,
            HelpMessage = 'The evaluation scriptblock'
        )]
        [scriptblock]$Until,
        [Parameter(
            Mandatory,
            ValueFromPipeline,
            HelpMessage = 'Pipeline input that will be processed'
        )]
        [ValidateNotNullOrEmpty()]
        [PSObject]$InputObject
    )

    Begin {
        Write-Verbose "[$((Get-Date).TimeOfDay) BEGIN  ] Starting $($MyInvocation.MyCommand)"
        Write-Verbose "[$((Get-Date).TimeOfDay) BEGIN  ] Running under PowerShell version $($PSVersionTable.PSVersion)"
        Write-Verbose "[$((Get-Date).TimeOfDay) BEGIN  ] Until : $Until"
    } #begin

    Process {
        if (. $Until) {
            $InputObject
            Write-Verbose "[$((Get-Date).TimeOfDay) PROCESS] Until requirement met. Ending $($MyInvocation.MyCommand)"
            Break
        }
        else {
            $InputObject
        }
    } #process

    End {
        #if we reach here, the Until condition was never met
        Write-Verbose "[$((Get-Date).TimeOfDay) END    ] Processed all input"
        Write-Verbose "[$((Get-Date).TimeOfDay) END    ] Ending $($MyInvocation.MyCommand)"
    } #end

} #close Select-Until
```

**Comment by mklement0** [comment]
@SteveL-MSFT, you can't use `break` that way in a function; not being able to do so is what motivated #3821.

E.g., `1, 2 | Select-Until { $true } -Verbose; 'foo'` never executes `'foo'`, because the `break` exits the entire script block (in the absence of an enclosing loop).

**Comment by MatejKafka** [comment]
@SteveL-MSFT I explicitly mention your proposed solution in the issue description, and it does not work, since `break` is not safe to use outside of a loop. Please re-evaluate.

**Comment by Jaykul** [comment]
Good grief. This should be on `Where-Object` to bring it up to parity with the `.Where` method. But it shouldn't have been closed with a incomplete example script.

**Comment by mklement0** [comment]
@Jaykul, re `Where-Object`, see:

* #13834

**Comment by Jaykul** [comment]
Yeah @Mkelement, not terribly helpful, I don't have time read 20 posts of a bunch of team and MVPs sniping each other over semantics on some _other_ failed attempt to explain that we need this.

**Comment by mklement0** [comment]
Yeah, @Jaykul, I agree: your response isn't terribly helpful.
Nobody actually cares what you think about incidental, ad-hominem aspects of a discussion that should be evaluated on its merits.

**Comment by iSazonov** [comment]
I want to remind you of the story. We tried to make changes to the Select-Object several times, and each time it caused prolonged dancing because we were breaking something all the time. This cmdlet is overly complex. As a result, a general consensus was reached - no more changes in this cmdlet, it is better to create a new one than to make changes to this one.
So I won't reopen the issue.

Nevertheless, the need (scenarios) described here exists and we should provide users with this opportunity one way or another.

I looked at #13834 what @mklement0 indicated. That proposal is interesting, but it was initially initiated as a conceptual discussion with many roads, which led to a dead end. Nevertheless, there were many useful thoughts.

Based on these two discussions, I propose to reach a simple compromise to provide users with this opportunity.
It could be a simple script parameter:
```powershell
... | Where-Object -FilterScript { ... } -StopScript { ... }
```
It may not be as beautiful as it was in these discussions, but it's very powerful and simple. If you like it, then open a new discussion - it has every chance of being implemented in this milestone.
